Caching OneDNN param for conv in TorchInductor

Guobing-Chen opened the pull request:
Add mechanism in TorchInductor to cache oneDNN related OP param which can be reused for these OPs in later invoking. Conv related OPs are updated to support this feature.

Two oneDNN param generation functions are added in pytorch to provide conv param generation based on input tensors and other parameters. They will return a param handler which been stored into a param cache sitting in TorchInductor generated sub-graph code. Conv related OPs in the generated sub-graph code will then query the cache to get the param and use it directly instead of initializing a conv param everytime been invoked.

TorchDynamo will guard the input shape change and invoke TorchInductor to re-generate new sub-graph code which will then include a new param cache for the new shapes.

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @VitalyFedyunin @mlazos @soumith @yanboliang @anijain2305 @chunyuan-w @desertfire

Guobing-Chen committed:
Caching OneDNN param for conv in TorchInductor

Add mechanism in TorchInductor to cache oneDNN related OP param which
can be reused for these OPs in later invoking. Conv related OPs are
updated to support this feature.

Guobing-Chen committed:
Fix lint issue

Guobing-Chen committed:
Update OP schema and add test cases

Guobing-Chen commented:
There are two opens need some discussion:
- The oneDNN param (a structure) is allocated in native space and hooked at the TorchInductor generated sub graph. As per TorchDynamo/TorchInductor's current design, the sub graph will keep alive there until the entire process exit, which means the allocated oneDNN param will never been freed until the process exit. The size of the param is very small but we can discuss more whether this mechanism is good.
- Can we assume the newly created conv/linear for post-op fusion like mkldnn::_convolution_pointwise/mkldnn::_convolution_pointwise.binary will only be used by TorchInductor path? If it is true, we can possibly remove some of the CHECK in these OPs as they have been already CHECKed during oneDNN param creation and TorchDynamo will guard the shape of input/weight/bias/output I suppose, which can assume the input tensors to the OP during runtime are following the same shape as when oneDNN param created.

Guobing-Chen commented:
@EikanWang , @jgong5 , pls help to have a review on this PR.

jgong5 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -334,6 +335,32 @@ def call(args):
+                else:
+                    conv_param = torch.ops.mkldnn._conv_param_generation(src, weight, bias, padding, stride, dilation,
+                                                                         groups, attr, scalars, algo)
+                    onednn_param_cache[op_id] = conv_param
```
How would these handles be deleted, e.g., wrapping it to a Python object and call delete from its `__del__` function?

Guobing-Chen replied:
Yes, that is exactly the open I raised in the PR comments that would like to discuss with you. In current TorchDynamo/TorchInductor design, the TorchInductor generated python wrapper will be parsed as a context and keeps alive until the entire process exit. Even we attach an exit mechanism seems it will not be triggered. Do we still need to add a mechanism to do that?

jgong5 replied:
I think the feature itself should do its job right, like having a way to get its handles disposed. Then it is wrapper's job to properly manage the lifecycle with this contract. The latter could be another design if it is missing now...

Guobing-Chen replied:
I've added mechanism for deleting these handles, and also added a config to control the param caching (with default to be False)

jgong5 reviewed (changes requested):
Can we add a config option to control the enabling of this feature?

Guobing-Chen commented on aten/src/ATen/native/mkldnn/RegisterMkldnnOpContextClass.cpp:
```diff
@@ -40,11 +40,15 @@ TORCH_LIBRARY(mkldnn, m) {
       "mkldnn::_linear_pointwise.binary(Tensor X, Tensor other, Tensor W, Tensor? B, str attr) -> Tensor Y"));
   m.def(TORCH_SELECTIVE_SCHEMA(
-      "mkldnn::_convolution_pointwise(Tensor X, Tensor W, Tensor? B, int[] padding, int[] stride, int[] dilation, int groups, str attr, Scalar?[] scalars, str? algorithm) -> Tensor Y"));
+      "mkldnn::_convolution_pointwise(Tensor X, Tensor W, Tensor? B, int[] padding, int[] stride, int[] dilation, int groups, str attr, Scalar?[] scalars, str? algorithm, int? param) -> Tensor Y"));
```
The int here in pytorch refers to int64_t, and in these conv kernels we will set the default to be 0 and fallback to normal path if it is truly 0. Suppose it should work. 

What is your potential concern about?

jgong5 replied:
Oh, if `int` means `int64_t` here, it looks fine.

Guobing-Chen committed:
Add support for mkldnn param deleting and set config to control param caching

Guobing-Chen commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -221,6 +256,197 @@ void _mkldnn_convolution_out (
+    conv_params_ptr = nullptr;
+  }
+
+  return (long) conv_params_ptr;
```
Updated to use int64_t instead.

jgong5 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -334,6 +335,59 @@ def call(args):
+                    else:
+                        return 0
+
+                def check_param(self, param_id):
+                    if param_id in self.p_cache:
+                        return True
+                    else:
+                        return False
```
Seems we can use `get_param(param_id) != 0` to check the existence in cache too?

Guobing-Chen replied:
I'd make the logic to be direct that check_param for judgement and get_param to obtain value.

jgong5 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -334,6 +335,59 @@ def call(args):
+
+                def get_param(self, param_id):
+                    if param_id in self.p_cache:
+                        return self.p_cache[param_id][0]
```
Can this work with multi-threaded case?

Guobing-Chen replied:
Good catch, seems I forgot to add multi-threading support which I originally plans to add after the single-thread case works. 

As for multi-threading, the major critical area is the cache write part, I will add threading lock for the cache writing code block and use double-checking to make the impact of lock to be only on cache miss case and only generate cache item once.

And for the cache clearing, as currently we hook it to the context deleting timepoint which should be per process instead of per thread, so I'd think we don't need to add lock for the param deleting logic.

jgong5 commented on torch/_inductor/ir.py:
```diff
@@ -3413,9 +3414,21 @@ def __init__(
+            if not wrapper.mkldnn_param_utils_gen:
+                wrapper.mkldnn_param_utils_gen = True
+                wrapper.write_mkldnn_param_gen_utils()
+            wrapper.writeline(f"{self.get_name()}_id = {random.getrandbits(32)}")
```
Do we really need randomize ids here? Would this break the persistent code cache hashing?

Guobing-Chen replied:
In my test env it always generate with same random id may be due to seed setting. Anyway, you are right we can use and ordered number instead of random. I've updated code accordingly.

Guobing-Chen committed:
Add multithreading support and refactor param generation code

Guobing-Chen commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -175,26 +179,87 @@ static inline at::MemoryFormat mkldnn_convolution_memory_format(int64_t dims, bo
-void _mkldnn_convolution_out (
+#define ONEDNN_PARAM_TYPE_CONV 0
+#define ONEDNN_PARAM_TYPE_LINEAR 1
+bool mkldnn_delete_param(int64_t param, int64_t param_type) {
```
The purpose of this function is to actually delete the param, instead of handle, so how about we update the function schema as `bool mkldnn_delete_param(int64_t param_handle, int64_t param_type)` ?

Guobing-Chen commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -175,26 +179,87 @@ static inline at::MemoryFormat mkldnn_convolution_memory_format(int64_t dims, bo
+  return true;
+}
+
+int64_t mkldnn_conv_param_generation(
```
The same reason as in delete case, the purpose of this function is to actually generate a new param, but return a param handle to the caller, so how about we make it as `int64_t mkldnn_create_conv_param_handle(...)`

Guobing-Chen committed:
Update naming of param related functional and parameters

Guobing-Chen committed:
Use ordered param id instead of random

jgong5 reviewed (approved):
LGTM

Guobing-Chen committed:
Update conv fusion op call in new mkldnn.py

Guobing-Chen committed:
Update test cases according to the new op and schema

jansel commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -175,26 +179,87 @@ static inline at::MemoryFormat mkldnn_convolution_memory_format(int64_t dims, bo
+        }
+        break;
+      default:
+        return false;
```
Should this throw an error?

Guobing-Chen replied:
Here we can only do real delete operation when we recognized the type, and do nothing (in default path) but return false to tell the caller that there is something wrong in his calling that he should deal with. This should be safe enough?

jansel commented on test/test_mkldnn_fusion.py:
```diff
@@ -302,6 +341,57 @@ def forward(self, x, other):
+                            x, other, mod.conv.weight, mod.conv.bias, mod.conv.padding, mod.conv.stride, mod.conv.dilation,
+                            mod.conv.groups, attr, None, unary_attr, [], None, conv_param
+                        )
+                        torch.mkldnn_delete_param(conv_param, 0)
```
IMO it would be cleaner if we could tie the garbage collection of these parameter objects to the python `conv_param` object, rather than exposing C++ memory management to python.  Python normally uses reference counting and you can run code in the destructor when the reference count goes to zero.

Guobing-Chen replied:
Yeah, here I just explictly use the delete_param function in test case to practice this function. In the real code in torch/_inductor/codegen/wrapper.py, these parameter objects are tied to mkldnn_param_cache python object, and will be auto deconstructed within the __del__. This mkldnn_delete_param will not be called directly by any other places explictly.

jansel replied:
This logic doesn't seem inductor-specific.  We should try to have a sane API at the `torch.ops.*` boundary.

jansel commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -353,6 +356,53 @@ def call(args):
                 )
             V.graph.sizevars.codegen(self.wrapper_call, V.graph.graph_inputs)
 
+    def write_mkldnn_param_gen_utils(self):
```
I don't think this caching based on ID will work with dynamic shapes.

Guobing-Chen replied:
Yes, here I assume TorchDynamo will guard the input shape change and invoke TorchInductor to re-generate new sub-graph code (a new python wrapper code, together with some included cpp kernels) which will then include a new param cache for the new shapes. Can we take this assumption? Or any suggestions?

Guobing-Chen committed:
Add support for dynamic shape

Update the caching key generation from just an OP id into OP_id +
multi-threading number + All input tensors' shape (but exclude
weight/bias as assume their shapes not change)

jgong5 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -7102,6 +7102,9 @@
 - func: can_cast(ScalarType from, ScalarType to) -> bool
   variants: function
 
+- func: mkldnn_delete_param(int param_handle, int param_type) -> bool
+  variants: function
```
This is something that cannot be registered and has to be declared here?

Guobing-Chen replied:
Due to that it is with no tensor params that will fail to dispatch when registered under mkldnn/CPU device, I've not changed to use torch.ops.mkldnn and CatchAll to get it work.

jgong5 commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -353,6 +356,63 @@ def call(args):
+                # Add op id and multi-threading number as part of key
+                key = [op_id, torch.get_num_threads()]
+                for arg in param_args[0]:
+                    # Add inputs tensors' shape as part of key, assume weight/bias shape will not change
```
Do we also have unit tests to cover the dynamic shape scenarios?

Guobing-Chen replied:
The problem is that seems TorchDynamo seems with problem under dynamic shape=True even with a simple model with only add/sub. Any suggested way to test this?

jgong5 replied:
One thing I realized was that we don't have to worry about dynamic shape handling with weight prepacked conv/linear since the weight prepacking needs the concrete input shapes to work. Also, for here, we don't have to check the input shapes and add them input the cache key if the input shapes are static, right?

Guobing-Chen committed:
Move _delete_param into torch.ops.mkldnn

jgong5 reviewed (commented):
LGTM in general but suggest to revise the dynamic shape handling to avoid checking input shapes when the input shapes are static.

jansel commented on aten/src/ATen/native/mkldnn/RegisterMkldnnOpContextClass.cpp:
```diff
@@ -40,11 +40,17 @@ TORCH_LIBRARY(mkldnn, m) {
+  m.def(TORCH_SELECTIVE_SCHEMA(
+      "mkldnn::_create_conv_param_binary(Tensor X, Tensor other, Tensor W, Tensor? B, int[] padding, int[] stride, int[] dilation, int groups, str binary_attr, Scalar? alpha, str? unary_attr, Scalar?[] unary_scalars, str? unary_algorithm) -> int param"));
+  m.def(TORCH_SELECTIVE_SCHEMA(
+      "mkldnn::_delete_param(int param_handle, int param_type) -> bool success"));
```
I don't think we should expose C++ memory management to python.  Suggest using `m.class()` to define a class that manages the lifecycle of parameters and relies on the Python garbage collector to figure out when to call  `_delete_param`.

I could also see a case for putting this cache in C++.
